Infrastructure: improve build type handling

SlySven opened the pull request:
#### Brief overview of PR changes/additions
Codifies the "type" of the Mudlet build to a single `enum class` instead of several `bool` flags and removes duplicate code to determine what value(s) they/it should have.

#### Motivation for adding to Mudlet
Simplification and removal of duplicate code. Also preparation wanted as part of solving problem with updating current Windows builds with the CI/CB produced "installer".

#### Other info (issues closed, discussion etc)
This is the first step in the resolution I am attempting for #7313.

SlySven committed:
Infrastructure: improve build type handling

#### Brief overview of PR changes/additions
Codifies the "type" of the Mudlet build to a single `enum class` instead
of several `bool` flags and removes duplicate code to determine what
value(s) they/it should have.

#### Motivation for adding to Mudlet
Simplification and removal of duplicate code. Also preparation wanted as
part of solving problem with updating current Windows builds with the CI/CB
produced "installer".

#### Other info (issues closed, discussion etc)
This is the first step in the resolution I am attempting for #7313.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/mudlet.cpp:
```diff
@@ -8591,3 +8605,24 @@ void mudlet::reattachOrphanedProfiles()
+    return mBuildType;
+}
+
+bool mudlet::isUpdateable() const
```
I would prefer a simpler `updateable()` instead, it makes it easier to read `if updateable then` rather than `if isUpdateable then`

SlySven replied:
Actually the naming means it could be confusing as to whether this is a build that can be updated or not OR whether this is an updatable build and an update is available/ready.

:thinking:

I'm going to change that from `isUpdateable` to `updateableBuild`...

vadi2 commented on src/mudlet.cpp:
```diff
@@ -211,11 +222,14 @@ void mudlet::init()
-    } else { // developmentVersion
+        break;
+    default:
         setWindowIcon(QIcon(qsl(":/icons/mudlet_dev_256px.png")));
```
Would non-updateable Linux distribution packages now show up as dev?

SlySven replied:
:thinking: I don't believe this PR makes a functional change in the code gated by `#if defined(INCLUDE_VARIABLE_SPLASH_SCREEN)` in `QImage mudlet::getSplashScreen()` but you may have a point about the icon used for Mudlet **here** for the code before or in this PR...

vadi2 commented on src/mudlet.h:
```diff
@@ -347,6 +346,15 @@ class mudlet : public QMainWindow, public Ui::main_window
+    bool isUpdateable() const;
+    QImage getSplashScreen();
+
+    // Try to avoid reading this value directly but instead access it via
```
In this case, a would it be better served as `private`?

SlySven replied:
Well this comment also applies to usage *within* the `mudlet` class - but yeah, with a getter there is no point in not making it `private`. Of course there are many other instances in Mudlet where classes provide getters and setters but still leave their member variables hanging-out for every~~one~~ *other class* to see/clobber...! :smirk:

vadi2 commented on src/TMap.cpp:
```diff
@@ -1883,34 +1883,31 @@ bool TMap::retrieveMapFileStats(QString profile, QString* latestFileName = nullp
+    if (fileVersion) {
+        *fileVersion = otherProfileVersion;
+    }
+
```
All three branches set *fileVersion = otherProfileVersion. This could be simplified:

```cpp
  if (fileVersion) {
      *fileVersion = otherProfileVersion;
  }

  if (otherProfileVersion > mDefaultVersion) {
      if (mudlet::self()->isUpdateable() || otherProfileVersion > mMaxVersion) {
          file.close();
          return true;
      }
  }
```

SlySven replied:
:thinking:

SlySven replied:
Okay - agreed - and I'll adopt it in a following commit.

vadi2 replied:
(Claude Opus 4.5 found this improvement and designed the solution, I cant claim credit)

vadi2 commented on src/mudlet.h:
```diff
@@ -137,6 +137,13 @@ class mudlet : public QMainWindow, public Ui::main_window
     mudlet();
     ~mudlet() override;
 
+    enum class BuildType {
+        Unknown = 0,
+        NotUpdateable, // Previously developmentVersion = true
+        Release, // Previously releaseVersion = true
+        PublicTest // Previously publicTestVersion = true
+    };
```
Going forward I think this would be more helpful for anyone working with the codebase:

```suggestion
    enum class BuildType {
      Unknown = 0,          // Uninitialized state (should never be used)
      NotUpdateable,        // Development or distribution package build
      Release,              // Official release build with auto-updater
      PublicTest            // Public test build (PTB) with auto-updater
  };
```

SlySven committed:
Fix: revise to remove 32-Bit windows update hint

This is now dead code and it contained a usage of the (now removed)
`(bool) mudlet::releaseVersion` flag which was borking just Windows builds.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: adopt suggestions from review

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

ZookaOnGit commented:
conflict, not building, switching to draft

SlySven committed:
Merge branch 'development' into Infrastructure_improveBuildTypeHandling

Conflicts resolved in:
* src/mudlet.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: rename function/method which was suggested in peer review

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: correct a Windows only build failure

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 reviewed (commented):
The bug discussed in https://github.com/Mudlet/Mudlet/pull/8582#discussion_r2147100389 is still present - Linux distribution packages without `INCLUDE_UPDATER` would show the dev icon instead of the release icon.

The issue is in the `mudlet::mudlet()` constructor. Currently:
```cpp
#if defined (INCLUDE_UPDATER)
    if (mAppBuild.isEmpty()) {
        mBuildType = BuildType::Release;
    } else {
        if (mAppBuild.startsWith(qsl("-ptb"))) {
            mBuildType = BuildType::PublicTest;
        } else {
            mBuildType = BuildType::NotUpdateable;
        }
    }
#else
    mBuildType = BuildType::NotUpdateable;  // Bug: ignores mAppBuild.isEmpty()
#endif
```

The `#else` branch unconditionally sets `NotUpdateable`, which then falls into the `default` case in the icon switch and shows the dev icon.

The fix is to check `mAppBuild.isEmpty()` even when `INCLUDE_UPDATER` is not defined:
```cpp
#if defined(INCLUDE_UPDATER)
    if (mAppBuild.isEmpty()) {
        mBuildType = BuildType::Release;
    } else if (mAppBuild.startsWith(qsl("-ptb"))) {
        mBuildType = BuildType::PublicTest;
    } else {
        mBuildType = BuildType::NotUpdateable;
    }
#else
    // Distro packages without updater: still use Release type for empty app-build.txt
    mBuildType = mAppBuild.isEmpty() ? BuildType::Release : BuildType::NotUpdateable;
#endif
```

This preserves the current behavior where `releaseVersion = mAppBuild.isEmpty()` is independent of whether the updater is compiled in.

vadi2 commented:
https://github.com/SlySven/Mudlet/pull/94

SlySven commented:
https://github.com/Mudlet/Mudlet/pull/8582#pullrequestreview-3544822735 isn't quite correct IMHO.
> This preserves the current behavior where releaseVersion = mAppBuild.isEmpty() is independent of whether the updater is compiled in.

In the context for the `mudlet::BuildType` `Release` DOES describe the behaviour needed by the code with regard to the functionality needed to perform **updates** - Linux packagers are ***expected*** to put something in `MUDLET_VERSION_BUILD` so that their builds are identified as something packaged into their distribution - for example if Debian ever returned to including Mudlet I would expect them to use a value like `-deb12u1`. However the functionality is convoluted in that the `WITH_VARIABLE_SPLASH_SCREEN` which was intended to make Mudlet always show the standard splash-screen would also be needed to make such a packaged build look like the "Release" one from us for the end-user - but for it to rely on their distribution specific updating mechanism rather than our (not in their case) built-in one.

I am actively considering changing `WITH_VARIABLE_SPLASH_SCREEN` to `WITH_BUILD_BRANDING` and using THAT to control both the (now only normally visible on the "About Mudlete" dialogue) splash screen AND the icon used for the Mudlet application itself (when NOT set to a `NO` value).

I am also recognising that I (well, we collectively) need to document all the `WITH_XXX` - and possibly other build "control" variables in the Wiki... :grinning:

vadi2 commented:
> Linux packagers are expected to put something in MUDLET_VERSION_BUILD

I haven't heard of this before, and I don't think any package managers do this right now. Not sure why should we start obligating them to do it either.

I recommend fixing the logic bug introduced in the PR.

SlySven commented:
> > Linux packagers are expected to put something in MUDLET_VERSION_BUILD
> 
> I haven't heard of this before, and I don't think any package managers do this right now. Not sure why should we start obligating them to do it either.
> 
> I recommend fixing the logic bug introduced in the PR.

The instruction has been there right from the blooming start :angry: - see 19a60d6c5ec05aa04fc6e692996fdfcf72b329b4 from 2017:
```diff
+# if you are distributing modified code, it would be useful if you
+# put something distinguishing into the MUDLET_VERSION_BUILD environment
+# variable to make identification of the used version simple
+# the qmake BUILD variable is NOT built-in
+BUILD = $$(MUDLET_VERSION_BUILD)
+isEmpty( BUILD ) {
+# Leave the value of the following empty for a release build
+# i.e. the line should be "BUILD =" without quotes
+BUILD = "-dev"
+}
```

vadi2 committed:
Merge branch 'development' into Infrastructure_improveBuildTypeHandling
